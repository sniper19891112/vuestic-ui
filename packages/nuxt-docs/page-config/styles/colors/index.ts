export default definePageConfig({
  blocks: [
    block.title("colors.title"),
    block.paragraph("colors.description"),

    block.subtitle("colors.theme.title"),
    block.paragraph("colors.theme.description"),
    block.paragraph("colors.theme.autoTextColor"),
    block.component("Theme", { hideTemplate: true }),
    block.component("ThemeExamples", { hideCode: true }),

    block.subtitle("colors.syntax.title"),
    block.paragraph("colors.syntax.description"),
    block.code("scheme"),
    block.paragraph("colors.syntax.colorsUsed"),
    block.example("CustomAlert", { hideTitle: true }),

    block.subtitle("colors.reactivity.subtitle"),
    block.paragraph("colors.reactivity.about"),
    block.paragraph("colors.reactivity.additional"),
  ],
});
