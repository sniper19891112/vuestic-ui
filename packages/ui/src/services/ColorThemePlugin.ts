import { mixins } from 'vue-class-component'
import Vue, { App, reactive } from 'vue'
import { makeContextablePropsMixin } from '../components/context-test/context-provide/ContextPlugin'

// declare module 'vue/types/vue' {
//   interface App {
//     // @deprecated, use getColor instead
//     $themes: Record<string, string> | undefined;
//   }
// }

// Most default color - fallback when nothing else is found.
const DEFAULT_COLOR = '#000000'

const defaultOptions = reactive({
  themes: {
    primary: '#23e066',
    secondary: '#002c85',
    success: '#40e583',
    info: '#2c82e0',
    danger: '#e34b4a',
    warning: '#ffc200',
    gray: '#b4b6b9',
    dark: '#34495e',
  },
})

export const THEME_NAMES = {
  DEFAULT: 'DEFAULT',
  CORPORATE: 'CORPORATE',
}

export const COLOR_THEMES = [
  {
    name: THEME_NAMES.DEFAULT,
    themes: {
      primary: '#40e583',
      secondary: '#002c85',
      success: '#40e583',
      info: '#2c82e0',
      danger: '#e34b4a',
      warning: '#ffc200',
      gray: '#babfc2',
      dark: '#34495e',
    },
  },
  {
    name: THEME_NAMES.CORPORATE,
    themes: {
      primary: '#6c7fee',
      secondary: '#6e7ff1',
      success: '#8ddc88',
      info: '#71baff',
      danger: '#f8706d',
      warning: '#ffd652',
      gray: '#8396a5',
      dark: '#34495e',
    },
  },
]

export const getDefaultOptions = () => defaultOptions

export const ColorThemePlugin = {
  install (app: App, options?: { themes?: Record<string, string> }) {
    if (options && options.themes) {
      defaultOptions.themes = { ...defaultOptions.themes, ...options.themes }
    }
    // (app as any).$options.$themes = defaultOptions.themes
    app.config.globalProperties.$themes = reactive(defaultOptions.themes)
  },
}

/**
 * Check if color is valid css color
 * Taken from https://stackoverflow.com/a/56266358/5783475
 * @param strColor
 */
export const isCssColor = (strColor: string): boolean => {
  const s = new Option().style
  s.color = strColor
  return s.color !== ''
}

// TODO fix App| any
export const getColor = ($app: App| any, prop: string, defaultColor: string = DEFAULT_COLOR): string => {
  const $themes = $app.$themes
  if ($themes && $themes[prop]) {
    return $themes[prop]
  }

  if (isCssColor(prop)) {
    return prop
  }

  return defaultColor
}

export class ColorThemeMixin extends mixins(
  makeContextablePropsMixin({
    color: {
      type: String,
    },
  }),
) {
  get colorComputed () {
    return getColor(this, this.c_color)
  }

  computeColor (prop: string, defaultColor?: string) {
    return getColor(this, prop, defaultColor)
  }

  setTheme (theme: Record<string, string>) {
    Object.assign(this.$themes, theme)
  }

  created () {
    this.hasColorThemeMixin = true
  }
}
