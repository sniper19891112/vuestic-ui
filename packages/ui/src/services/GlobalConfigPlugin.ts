import { ref, inject, App } from 'vue'

import { getDefaultConfig } from '../components/vuestic-components/va-config/config-default'
import { merge } from 'lodash'
import { colorThemes } from './color-theme/color-theme-presets'
import { ColorConfig } from './color-theme/color-config'
import { ComponentConfig } from './component-config/component-config'

// export type GlobalConfig = Record<string, Record<string, any> | undefined> & { theme?: ColorTheme };

export type GlobalConfig = {
  colors?: ColorConfig,
  icons?: IconConfig
  components?: ComponentConfig
} & { theme?: ColorConfig };

type Updater = (config: GlobalConfig) => GlobalConfig;

/**
 * The global configuration reference
 */
const globalConfigRef = ref({
  theme: colorThemes.default as ColorConfig,
  components: {
    // TODO Component configs should go there.
  },
  ...getDefaultConfig(),
}) as Record<string, any>

export const GLOBAL_CONFIG = Symbol('GLOBAL_CONFIG')

export function useGlobalConfig () {
  const globalConfig = inject(GLOBAL_CONFIG, {} as any)

  return {
    getGlobalConfig: globalConfig?.get,
    setGlobalConfig: globalConfig?.set,
  }
}

const setGlobalConfig = (updater: GlobalConfig | Updater): void => {
  if (typeof updater === 'function') {
    globalConfigRef.value = merge(
      globalConfigRef.value,
      updater(globalConfigRef.value),
    )
  } else {
    globalConfigRef.value = merge(globalConfigRef.value, updater)
  }
}

const getGlobalConfig = (): GlobalConfig => globalConfigRef.value

/**
 * Plugin provides global config to Vue component through prototype
 */
const GlobalConfigPlugin = {
  install (app: App, options?: GlobalConfig) {
    if (options) { setGlobalConfig(options) }

    const config = { get: getGlobalConfig, set: setGlobalConfig }

    app.provide(GLOBAL_CONFIG, config)
  },
}

export default GlobalConfigPlugin
