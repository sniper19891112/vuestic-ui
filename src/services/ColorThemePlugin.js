const getDefaultOptions = () => ({
  themes: {
    primary: '#40e583',
    secondary: '#002c85',
    success: '#40e583',
    info: '#2c82e0',
    danger: '#e34b4a',
    warning: '#ffc200',
    gray: '#babfc2',
    dark: '#34495e',
  },
})

export const ColorThemePlugin = {
  install (Vue, options) {
    const defaultOptions = getDefaultOptions()

    if (options && options.themes) {
      Object.assign(defaultOptions.themes, options.themes)
    }

    Vue.prototype.$themes = defaultOptions.themes

    /* eslint-disable no-new */
    // This line is just to make themes reactive
    new Vue({ data: { themes: Vue.prototype.$themes } })
  },
}

export const ColorThemeMixin = {
  data () {
    return {
      colorThemeDefault: 'primary',
      colorDefault: '#000000',
    }
  },
  props: {
    color: {
      type: String,
    },
  },
  computed: {
    _isEnableColorTheme () {
      return this.$themes
    },
    // This allows a multitude of defaults.
    // theme color => color => theme default => hard default
    colorComputed () {
      return this.computeColor(this.color)
    },
  },
  methods: {
    computeColor (prop) {
      if (this._isEnableColorTheme) {
        if (this.$themes && this.$themes[prop]) {
          return this.$themes[prop]
        }
        if (this.$themes && this.$themes[this.colorThemeDefault]) {
          return this.$themes[this.colorThemeDefault]
        }
      }

      if (prop) {
        return prop
      }

      return this.colorDefault
    },
  },
}
