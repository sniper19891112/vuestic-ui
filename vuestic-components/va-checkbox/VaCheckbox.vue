<template>
  <div
    class="va-checkbox"
    :class="computedClass"
  >
    <div
      class="va-checkbox__input-container"
      @click="toggleSelection()"
      @mousedown="onMouseDown"
      @mouseup="onMouseUp"
    >
      <div
        class="va-checkbox__square"
        :class="{'active': value}"
      >
        <input
          :id="id"
          readonly
          @focus="onFocus"
          @blur="isKeyboardFocused = false"
          class="va-checkbox__input"
          @keypress.prevent="toggleSelection()"
          :disabled="disabled"
        />
        <va-icon icon="ion ion-md-checkmark va-checkbox__icon-selected"/>
      </div>
      <div class="va-checkbox__label-text">
        <slot name="label">
          {{ label }}
        </slot>
      </div>
    </div>
    <div class="va-checkbox__error-message-container" v-if="showError">
      <div
        class="va-checkbox__error-message"
        v-for="(error, index) in computedErrorMessages"
        :key="index"
      >
        {{ error }}
      </div>
    </div>
  </div>
</template>

<script>
import VaIcon from '../va-icon/VaIcon'

export default {
  name: 'va-checkbox',
  components: { VaIcon },
  props: {
    label: String,
    value: {
      type: Boolean,
      required: true,
    },
    id: {
      type: String,
    },
    disabled: {
      type: Boolean,
      default: false,
    },
    readonly: {
      type: Boolean,
      default: false,
    },
    errorMessages: {
      type: [String, Array],
      default: () => [],
    },
    errorCount: {
      type: Number,
      default: 1,
    },
    name: String,
    error: {
      type: Boolean,
      default: false,
    },
  },
  data () {
    return {
      isKeyboardFocused: false,
      hasMouseDown: false,
    }
  },
  computed: {
    computedClass () {
      return {
        'va-checkbox--selected': this.value,
        'va-checkbox--readonly': this.readonly,
        'va-checkbox--disabled': this.disabled,
        'va-checkbox--error': this.showError,
        'va-checkbox--on-keyboard-focus': this.isKeyboardFocused,
      }
    },
    computedErrorMessages () {
      const isArray = Array.isArray(this.errorMessages)
      const errorMessages = isArray ? this.errorMessages : [this.errorMessages]
      return errorMessages.slice(0, this.errorCount)
    },
    valueProxy: {
      set (value) {
        this.$emit('input', value)
      },
      get () {
        return this.value
      },
    },
    showError () {
      // We make error active, if the error-message is not empty and checkbox is not disabled
      if (!this.disabled) {
        if (!(this.errorMessages.length === 0) || this.error) {
          return true
        }
      }
      return false
    },
  },
  methods: {
    onFocus (e) {
      if (this.hasMouseDown) {
        return
      }
      this.isKeyboardFocused = true
    },
    onMouseDown (e) {
      this.hasMouseDown = true
      this.$emit('mousedown', e)
    },
    onMouseUp (e) {
      this.hasMouseDown = false
      this.$emit('mouseup', e)
    },
    toggleSelection () {
      if (this.readonly) {
        return
      }
      if (this.disabled) {
        return
      }
      this.valueProxy = !this.valueProxy
    },
  },
}
</script>

<style lang="scss">
@import "../../vuestic-sass/resources/resources";

.va-checkbox {
  margin-bottom: $checkbox-between-items-margin;
  display: flex;
  flex-direction: column;

  &__input-container {
    align-items: center;
    display: flex;
    cursor: pointer;

    @at-root {
      .va-checkbox--disabled  & {
        @include va-disabled();
      }

      .va-checkbox--readonly & {
        cursor: initial;
      }

      .va-checkbox--disabled & {
        cursor: default;
      }
    }
  }

  #{&}__square {
    @include flex-center();
    width: 2rem;
    height: 2rem;
    position: relative;
    flex: 0 0 2rem;
    @at-root {
      .va-checkbox--on-keyboard-focus#{&} {
        background-color: $light-gray;
        transition: all, 0.6s, ease-in;
        border-radius: 5rem;
      }
    }
  }

  #{&}__input {
    height: 1.375rem;
    width: 1.375rem;
    cursor: inherit;
    color: $white;
    background-color: $white;
    border: solid 0.125rem $gray-light;
    border-radius: 0.25rem;

    &:focus {
      outline: none;
    }

    @at-root {
      .va-checkbox--selected#{&} {
        display: flex;
        justify-content: center;
        align-items: center;
        height: 1.4rem;
        width: 1.4rem;
        color: $white;
        background-color: $vue-green;
        border: 0;
      }

      .va-checkbox--error#{&} {
        border-color: $theme-red;
      }
    }
  }

  #{&}__label-text {
    display: inline-block;
    position: relative;
    margin-left: 0.25rem;
    @at-root {
      .va-checkbox--error#{&} {
        color: $theme-red;
      }
    }
  }

  &__error-message {
    vertical-align: middle;
    color: $theme-red;
    font-size: $font-size-mini;
  }

  &__icon-selected {
    pointer-events: none;
    position: absolute;
    color: $white;
  }

  &__error-message-container {
    flex: 0 0 100%;
    margin-left: 0.3rem; // To fit with checkbox.
  }

  &__label-container {
    margin-left: 2rem;
  }

  &__content {
    flex-direction: row;
  }
}
</style>
