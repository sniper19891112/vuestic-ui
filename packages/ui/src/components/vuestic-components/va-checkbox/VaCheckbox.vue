<template>
  <va-input-wrapper
    class="va-checkbox"
    :class="computedClass"
    :disabled="c_disabled"
    :success="c_success"
    :messages="c_messages"
    :error="computedError"
    :error-messages="computedErrorMessages"
    :error-count="errorCount"
  >
    <div
      class="va-checkbox__input-container"
      @click="onWrapperClick()"
      @mousedown="hasMouseDown = true"
      @mouseup="hasMouseDown = false"
      tabindex="-1"
      @blur="onBlur"
      ref="container"
    >
      <div
        class="va-checkbox__square"
        :style="inputStyle"
      >
        <input
          ref="input"
          :id="id"
          :name="name"
          type="checkbox"
          role="checkbox"
          readonly
          @focus="onFocus"
          @blur="onBlur($event)"
          class="va-checkbox__input"
          @keypress.prevent="toggleSelection()"
          :disabled="c_disabled"
          :indeterminate="c_indeterminate"
        >
        <va-icon
          class="va-checkbox__icon"
          :name="computedIconName"
          size="20px"
        />
      </div>
      <div
        class="va-checkbox__label"
        :style="labelStyle"
        ref="label"
        tabindex="-1"
        @blur="onBlur"
      >
        <slot name="label">
          {{ c_label }}
        </slot>
      </div>
    </div>
  </va-input-wrapper>
</template>

<script>
import VaIcon from '../va-icon/VaIcon'
import { SelectableMixin } from '../../vuestic-mixins/SelectableMixin/SelectableMixin'
import { makeContextablePropsMixin } from '../../context-test/context-provide/ContextPlugin'
import { getColor } from '../../../services/ColorThemePlugin'
import VaInputWrapper from '../va-input/VaInputWrapper'

export default {
  name: 'VaCheckbox',
  components: { VaInputWrapper, VaIcon },
  mixins: [
    SelectableMixin,
    makeContextablePropsMixin({
      value: { type: [Boolean, Array, String, Object], default: false },
      checkedIcon: { type: String, default: 'check' },
      indeterminateIcon: { type: String, default: 'remove' },
      color: { type: String, default: 'primary' },
    }),
  ],
  computed: {
    computedClass () {
      return {
        'va-checkbox--selected': this.isChecked,
        'va-checkbox--readonly': this.c_readonly,
        'va-checkbox--disabled': this.c_disabled,
        'va-checkbox--indeterminate': this.c_indeterminate,
        'va-checkbox--error': this.computedError,
        'va-checkbox--left-label': this.c_leftLabel,
        'va-checkbox--on-keyboard-focus': this.isKeyboardFocused,
      }
    },
    labelStyle () {
      return {
        color: this.computedError ? getColor(this, 'danger') : '',
        padding: !this.c_label
          ? ''
          : this.c_leftLabel
            ? '0 0.25rem 0 0'
            : '0 0 0 0.25rem',
      }
    },
    inputStyle () {
      return this.computedError
        ? (this.isChecked || this.isIndeterminate)
          ? { background: this.colorComputed, borderColor: getColor(this, 'danger') }
          : { borderColor: getColor(this, 'danger') }
        : (this.isChecked || this.isIndeterminate)
          ? { background: this.colorComputed, borderColor: this.colorComputed }
          : {}
    },
    computedIconName () {
      return (this.c_indeterminate && this.isIndeterminate)
        ? this.c_indeterminateIcon
        : this.c_checkedIcon
    },
  },
}
</script>

<style lang="scss" scoped>
@import "../../vuestic-sass/resources/resources";

.va-checkbox {
  display: flex;
  flex-direction: column;
  justify-content: center;

  &__input-container {
    align-items: center;
    display: flex;
    padding: 0 0.3rem;
    cursor: pointer;

    @at-root {
      .va-checkbox--disabled & {
        @include va-disabled();

        cursor: default;
      }

      .va-checkbox--readonly & {
        cursor: initial;
      }

      .va-checkbox--left-label & {
        flex-direction: row-reverse;
      }
    }
  }

  #{&}__square {
    @include flex-center();

    width: 1.35rem;
    min-width: 1.35rem;
    height: 1.35rem;
    position: relative;
    background-color: $white;
    border: solid 0.125rem $gray-light;
    border-radius: 0.25rem;

    @at-root {
      .va-checkbox--on-keyboard-focus#{&} {
        transition: all, 0.6s, ease-in;
        box-shadow: 0 0 0.5rem 0 rgba(0, 0, 0, 0.3);
      }
    }
  }

  &__input {
    opacity: 0;
    width: 0;
    height: 0;
  }

  &__label {
    display: inline-block;
    position: relative;
  }

  &__icon {
    pointer-events: none;
    position: absolute;
    color: transparent;
  }

  &--selected {
    .va-checkbox {
      &__icon {
        color: $white;
      }
    }
  }

  &--indeterminate {
    .va-checkbox {
      &__icon {
        color: $white;
      }
    }
  }
}

</style>
