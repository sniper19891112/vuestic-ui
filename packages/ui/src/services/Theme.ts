import { GlobalConfig, useGlobalConfig } from './GlobalConfigPlugin'

// Most default color - fallback when nothing else is found.
export const DEFAULT_COLOR = '#000000'

// This object is intended to be mutable, so that other services can observe it.
export const DEFAULT_THEME = {
  primary: '#23e066',
  secondary: '#002c85',
  success: '#40e583',
  info: '#2c82e0',
  danger: '#e34b4a',
  warning: '#ffc200',
  gray: '#b4b6b9',
  dark: '#34495e',
}

export const THEME_NAMES = {
  DEFAULT: 'DEFAULT',
  CORPORATE: 'CORPORATE',
}

export const COLOR_THEMES = [
  {
    name: THEME_NAMES.DEFAULT,
    themes: {
      primary: '#40e583',
      secondary: '#002c85',
      success: '#40e583',
      info: '#2c82e0',
      danger: '#e34b4a',
      warning: '#ffc200',
      gray: '#babfc2',
      dark: '#34495e',
    },
  },
  {
    name: THEME_NAMES.CORPORATE,
    themes: {
      primary: '#6c7fee',
      secondary: '#6e7ff1',
      success: '#8ddc88',
      info: '#71baff',
      danger: '#f8706d',
      warning: '#ffd652',
      gray: '#8396a5',
      dark: '#34495e',
    },
  },
]

export const useTheme = () => {
  const { setGlobalConfig, getGlobalConfig } = useGlobalConfig()

  if (!setGlobalConfig && !getGlobalConfig) {
    return
  }

  const setTheme = (theme: Record<string, string>): void => {
    setGlobalConfig((config: GlobalConfig) => ({
      ...config,
      theme: { ...config.theme, ...theme },
    }))
  }

  const getTheme = (): Record<string, string> | undefined => {
    return getGlobalConfig().theme
  }

  return {
    setTheme,
    getTheme,
  }
}
