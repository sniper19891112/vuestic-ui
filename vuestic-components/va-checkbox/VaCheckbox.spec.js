import Vue from 'vue'
import { shallowMount } from '@vue/test-utils'
import VaCheckbox from './VaCheckbox'

import { ColorPlugin } from './../../../services/colors'
Vue.use(ColorPlugin)

describe('VaCheckbox', () => {
  it('default', () => {
    const wrapper = shallowMount(VaCheckbox, {
      propsData: { value: false },
    })
    expect(wrapper.html()).toMatchSnapshot()
  })
  it('true value', () => {
    const wrapper = shallowMount(VaCheckbox, {
      propsData: { value: true },
    })
    expect(wrapper.html()).toMatchSnapshot()
  })
})
