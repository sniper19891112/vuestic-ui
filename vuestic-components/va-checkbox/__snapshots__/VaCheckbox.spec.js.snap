// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`VaCheckbox default 1`] = `
<div class="va-checkbox">
  <div class="va-checkbox__input-container">
    <div class="va-checkbox__square"><input readonly="readonly" class="va-checkbox__input">
      <va-icon-stub icon="ion ion-md-checkmark va-checkbox__icon-selected"></va-icon-stub>
    </div>
    <div class="va-checkbox__label-text">

    </div>
  </div>
  <!---->
</div>
`;

exports[`VaCheckbox true value 1`] = `
<div class="va-checkbox va-checkbox--selected">
  <div class="va-checkbox__input-container">
    <div class="va-checkbox__square active"><input readonly="readonly" class="va-checkbox__input">
      <va-icon-stub icon="ion ion-md-checkmark va-checkbox__icon-selected"></va-icon-stub>
    </div>
    <div class="va-checkbox__label-text">

    </div>
  </div>
  <!---->
</div>
`;
